storage: absorb RaftCommand into ReplicatedProposalData

tbg opened the pull request:
Remove the `storagebase.RaftCommand` proto by moving its fields onto
`ReplicatedProposalData` while preserving the tag numbers. Use that message
instead of `RaftCommand` throughout, including on the wire (since the tag
numbers were preserved, this does not require any special handling).

This in preparation for a follow-up change which adds an experimental switch to
use proposer-evaluated KV (#6166).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10200)

<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1720 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwt6PKsTPNPdkF1WPl:-KUwt6PKsTPNPdkF1WPm:b-fn7v7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/replica.go#L1720)):*
> ```Go
> 	// since evaluating a proposal is expensive (at least under proposer-
> 	// evaluated KV).
> 	var pd ProposalData
> ```

drive-by: is there any reason you're not using a composite literal here?

---

*[pkg/storage/replica.go, line 2834 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwt_sMHF4URCtkbFiP:-KUwt_sMHF4URCtkbFiQ:b-ttlbqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/replica.go#L2834)):*
> ```Go
> 
> 	// TODO(tschottdorf): not all commands have a BatchRequest (for example,
> 	// empty appends). Be more careful with this in proposer-eval'ed KV.
> ```

is the intention that some day you won't blindly allocate here? for now, perhaps it's better to make the field not-nullable and put the TODO there?

---

*[pkg/storage/storagebase/proposer_kv.proto, line 58 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwslnywMMjF1irP8Pd:-KUwslnywMMjF1irP8Pe:b-bg7jp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L58)):*
> ```protobuf
> message ReplicatedProposalData {
>   // ======================================
>   // Begin of what was formerly RaftCommand
> ```

either s/of // or s/Begin/Beginning/

---

*[pkg/storage/storagebase/proposer_kv.proto, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwtBHPZTpTYAxnIrMU:-KUwtBHPZTpTYAxnIrMV:b-l9os6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L65)):*
> ```protobuf
>       (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.RangeID"];
>   optional roachpb.ReplicaDescriptor origin_replica = 2 [(gogoproto.nullable) = false];
>   optional roachpb.BatchRequest cmd = 3;
> ```

how come this is nullable now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-:-KUwuJyXWonEHzNAscoE:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwtBHPZTpTYAxnIrMU:-KUxIcfjHqTayFaM2MIs:bs6fh31) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L65)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come this is nullable now?
</details>
Probably because it will be optional when we add the write batch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-:-KUxIuf-6eII47yjW9RW:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUyzGZy-qfgThgLcNaV:-KUyzGZy-qfgThgLcNaW:bky3364) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L99)):*
> ```protobuf
> 
>   // Whether to block concurrent readers while processing the proposal data.
>   optional bool block_reads = 5 [(gogoproto.nullable) = false];
> ```

Is this protobuf written over the network yet or is it purely in memory? (i.e. do we need to be careful about upgrades for this?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-:-KUyzSpE54GXDiigD_BQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1720 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwt6PKsTPNPdkF1WPl:-KUzjFlCWAzURcfM5MUo:b-2z1sul) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/replica.go#L1720)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> drive-by: is there any reason you're not using a composite literal here?
</details>
I have follow-up work in the pipeline for which this is better the way it is.

---

*[pkg/storage/replica.go, line 2834 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwt_sMHF4URCtkbFiP:-KUzjVDpqs__MU5Z34p7:b-nhtnqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/replica.go#L2834)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is the intention that some day you won't blindly allocate here? for now, perhaps it's better to make the field not-nullable and put the TODO there?
</details>
Empty appends are rare. Previous code below this line was mostly oblivious to the fact that sometimes it would run on a zero `BatchRequest`, and I wanted this PR to be devoid of actual semantic changes even though it's tempting. Will clean that up in the next ones (where I'll also introduce a switch that makes it always empty when run with proposer-eval'ed KV).

---

*[pkg/storage/storagebase/proposer_kv.proto, line 58 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwslnywMMjF1irP8Pd:-KUzkPSUNMEaVk7ZbrFB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L58)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> either s/of // or s/Begin/Beginning/
</details>
Done.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUwtBHPZTpTYAxnIrMU:-KUzkQNozdvTIodO6c_a:bb77ucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L65)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Probably because it will be optional when we add the write batch.
</details>
Yep.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUyzGZy-qfgThgLcNaV:-KUzkRYKzsuqxtRSVjyC:b-hnjt8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L99)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Is this protobuf written over the network yet or is it purely in memory? (i.e. do we need to be careful about upgrades for this?)
</details>
It's written over the network and to all Raft log. The fields in this section are all unused (future PR will introduce experimental switch to use them), so *this merge* should be OK (as older version simply ignore all fields). But you're right, we might have to be careful to change field types or tags here later (or a version that knows these tags exist might stumble over their type changing even for the zero value)? I'll hold off on the merge until you've had a chance to respond. The intention for those fields is definitely to be the wild west for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200#-KUyzGZy-qfgThgLcNaV:-KVL-oq26i72QoE5CoXu:b-tmujvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/529d0e2ae5e1f0184017bc418cdf448a12887756/pkg/storage/storagebase/proposer_kv.proto#L99)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> It's written over the network and to all Raft log. The fields in this section are all unused (future PR will introduce experimental switch to use them), so *this merge* should be OK (as older version simply ignore all fields). But you're right, we might have to be careful to change field types or tags here later (or a version that knows these tags exist might stumble over their type changing even for the zero value)? I'll hold off on the merge until you've had a chance to respond. The intention for those fields is definitely to be the wild west for now.
</details>
Sorry, I know that the old RaftCommand was written to the logs and the new ReplicatedProposalData will be; my question is whether the old ReplicatedProposalData was ever written to disk or network. (and it sounds like the answer is no, but it's not 100% clear)

I'm uncomfortable with allowing "wild west" changes to protobufs, especially ones that are this critical. I think that as long as the changes are limited to *nullable* fields that were never used then type changes would be OK, but it would be better to follow the regular protobuf rules (e.g. no type changes, just remove/add fields). We don't have to be as strict about it since these fields are not used - we don't need to preserve compatibility for older data that may have been written; we just don't want any inadvertently-written zero values to cause errors. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10200)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I've changed commentary so that it's clear that these fields are going to see cleanup and that they follow regular rules, and moved them to a high range of tags so that they don't make low tag numbers unavailable.

tschottdorf commented:
Oh, and to answer your question, no. The old one wasn't.

tbg committed:
storage: absorb RaftCommand into ReplicatedProposalData

Remove the `storagebase.RaftCommand` proto by moving its fields onto
`ReplicatedProposalData` while preserving the tag numbers. Use that message
instead of `RaftCommand` throughout, including on the wire (since the tag
numbers were preserved, this does not require any special handling).

This in preparation for a follow-up change which adds an experimental switch to
use proposer-evaluated KV (#6166).
